### DISCUSSION TITLE
How to create conditional breakpoints

### DISCUSSION THREAD

**Discussion by CharlesARoy** [mfussenegger/nvim-dap#670] [Q&A] [answered]
From issues such as [this](https://github.com/mfussenegger/nvim-dap/issues/289) it seems like nvim-dap supports creating conditional breakpoints, but I can't find any documentation on how. 

When I look at the available commands for DAP, I see DapToggleBreakpoint but nothing for conditional breakpoints.

![image](https://user-images.githubusercontent.com/46253001/188762282-3e494b75-67ff-4ece-b455-a22e75688c71.png)

Any help is appreciated!

**Comment by mfussenegger** [comment, marked as answer]
See `:help dap.toggle_breakpoint`

**Reply by CharlesARoy** [reply]
Thanks @mfussenegger! I don't mean to be obtuse but I'm still a little confused about two things:

1. I thought that a conditional breakpoint was a breakpoint that triggered when some condition was met _irrespective of line/location in the script_. E.g. the script stops if the variable i equals 5 anywhere in the script. Does such functionality exist?
2. I tried to use a conditional breakpoint in a simple Python script but it didn't seem to work as expected. 

I ran this command in Nvim: `:lua require('dap').toggle_breakpoint(i==5)` which created a breakpoint as seen here:
![image](https://user-images.githubusercontent.com/46253001/189178723-484c29b7-730b-4e5e-bfe8-aed906cecef1.png)

Then when I run the script, it stopped at the breakpoint even though the variable i was not equal to 5. Is that normal?
![image](https://user-images.githubusercontent.com/46253001/189178832-0870fa83-8a4b-4f43-834f-156122dfde81.png)

**Reply by mfussenegger** [reply]
1) Conditional breakpoints are still depending on the line where you place them. Variables can be in different scopes and it needs to know which ones you're referring to.

2) works fine for me. Make sure the condition is a string. `:lua require('dap').toggle_breakpoint('i==5')` otherwise `i == 5` gets treated as lua and is evaluated to `false` (assuming `i` doesn't exist as global variable in the lua interpreter)

**Reply by leosmaia21** [reply]
Hello, in codelldb how can i set a break with a function, like break if i == strlen(s)?

**Comment by fidelski** [comment]
Thanks a lot for this functionality! It's great to have that in neovim! I am still struggling to make conditional breakpoints work. I can specify the condition, it is shown in the dap-ui Breakpoints window and repl does not report any error. Still the debugger stops at the breakpoint even though the condition is not met.

So e.g. I set the condition to ```'i == 38'``` in a for-loop over an incrementing ```i``` but the debugger stops there for every value of ```i```.

Another example, I set the condition to ```'result == false'``` on this code:

```cpp
result = validate_checksum(fops, fp, header.checksum, header.size);
if (result) { // <- Breakpoint set here
     fops.seek(fp, next_pos, SEEK_SET);
}
```
But the debugger stops every time the breakpoint is reached, no matter what ```result``` is.

I tried to set the condition as string with ```'result == false'``` and without quotes as ```result == false``` but no difference, it doesn't work either way.

![quoted-condition](https://github.com/mfussenegger/nvim-dap/assets/858144/a69590ed-4bd3-4551-818a-1cefc1d91f99)

![no-quotes-condition](https://github.com/mfussenegger/nvim-dap/assets/858144/7c1c9399-ae04-4242-a954-7521cba7d170)


What am I missing?

**Comment by sleeptightAnsiC** [comment]
Take a look at code from PR https://github.com/mfussenegger/nvim-dap/pull/1325; it enhances `DapToggleBreakpoint` command with optional condition.

**Reply by fidelski** [reply]
Thanks for the hint! I indeed didn't catch up with the latest changes. I can confirm that my problem does not exist anymore. I can now successfully create conditional breakpoints that correctly get hit only when the condition is true.

**Comment by johan-lejdung** [comment]
Here are the keymaps I use specifically for breakpoints, it could perhaps help someone:

Lazyvim
```lua
return {
  "mfussenegger/nvim-dap",
  dependencies = {
    ...
  },
  keys = {
    ....,
    {
      "<leader>db",
      function()
        require("dap").toggle_breakpoint()
      end,
      desc = "Debug: Toggle Breakpoint",
    },
    {
      "<leader>dB",
      function()
        local condition = vim.fn.input("Breakpoint condition (optional): ")
        local hit_condition = vim.fn.input("Hit count (optional): ")

        -- Convert empty strings to nil
        condition = condition ~= "" and condition or nil
        hit_condition = hit_condition ~= "" and hit_condition or nil

        require("dap").toggle_breakpoint(condition, hit_condition)
      end,
      desc = "Debug: Toggle Advanced Breakpoint",
    },
  },
  ...
}
```
